mypy --strict error when importing numpy.typing

`mypy --strict` disables `implicit_reexport`. Consequently mypy throws an error when trying to import `ArrayLike` and `DTypeLike` from `numpy.typing`.

### Reproducing code example:

```python
from numpy.typing import ArrayLike
```

### Error message:

```
$ mypy --strict foo.py
foo.py:1: error: Module 'numpy.typing' has no attribute 'ArrayLike'
Found 1 error in 1 file (checked 1 source file)
```

### NumPy/Python version information:

```
$ python -c 'import sys, numpy; print(numpy.__version__, sys.version)'
1.20.0rc2 3.8.6 (default, Dec 18 2020, 13:50:46) 
[GCC 10.2.0]

$ mypy --version
mypy 0.790
```

### Possible workarounds

I guess that the reason why this happens is than `__all__` is not set when on `TYPE_CHECKING` mode:
https://github.com/numpy/numpy/blob/af8e67c4fd368bae710b01f32611b10dfab02b61/numpy/typing/__init__.py#L156-L165

If setting `__all__` is not possible, a workaround might be to "explicitly reexport" the objects using something this:

``` python
from ._array_like import ArrayLike as ArrayLike
from ._dtype_like import DTypeLike as DTypeLike
```

It might also be possible to enable implicit reexports for this specific module using the numpy mypy plugin.

--- a/numpy/typing/__init__.py
+++ b/numpy/typing/__init__.py
@@ -295,9 +295,9 @@ from ._scalars import (
     _VoidLike,
 )
 from ._shape import _Shape, _ShapeLike
-from ._dtype_like import _SupportsDType, _VoidDTypeLike, DTypeLike
+from ._dtype_like import _SupportsDType, _VoidDTypeLike, DTypeLike as DTypeLike
 from ._array_like import (
-    ArrayLike,
+    ArrayLike as ArrayLike,
     _ArrayLike,
     _NestedSequence,
     _SupportsArray,

